File: gin/internal/model/tournaments/tournament.go
package model_tournaments

import (
	"time"

	"gorm.io/gorm"
	"cards_project/internal/types"
	"fmt"
	"encoding/json"
)

var _ = json.RawMessage{}
type TournamentStatusType string
const (
	TournamentStatusType_Draft TournamentStatusType = "Draft"
	TournamentStatusType_Registration TournamentStatusType = "Registration"
	TournamentStatusType_Ongoing TournamentStatusType = "Ongoing"
	TournamentStatusType_Completed TournamentStatusType = "Completed"
	TournamentStatusType_Cancelled TournamentStatusType = "Cancelled"
)

type TournamentFormatType string
const (
	TournamentFormatType_Standard TournamentFormatType = "Standard"
	TournamentFormatType_Extended TournamentFormatType = "Extended"
	TournamentFormatType_Legacy TournamentFormatType = "Legacy"
	TournamentFormatType_Vintage TournamentFormatType = "Vintage"
	TournamentFormatType_Commander TournamentFormatType = "Commander"
	TournamentFormatType_Draft TournamentFormatType = "Draft"
)

type TournamentTournamentTypeType string
const (
	TournamentTournamentTypeType_Swiss TournamentTournamentTypeType = "Swiss"
	TournamentTournamentTypeType_SingleElimination TournamentTournamentTypeType = "SingleElimination"
	TournamentTournamentTypeType_DoubleElimination TournamentTournamentTypeType = "DoubleElimination"
	TournamentTournamentTypeType_RoundRobin TournamentTournamentTypeType = "RoundRobin"
)

// TournamentCreateRequest is the POST body.
type TournamentCreateRequest struct {
	PublicId string `json:"public_id" binding:"required"`
	Name string `json:"name" binding:"required"`
	Description *string `json:"description"`
	Status TournamentStatusType `json:"status" binding:"required"`
	BracketData *json.RawMessage `json:"bracket_data"`
	Format TournamentFormatType `json:"format" binding:"required"`
	TournamentType TournamentTournamentTypeType `json:"tournament_type" binding:"required"`
	MaxPlayers int `json:"max_players" binding:"min=2,max=512"`
	EntryFee types.Decimal `json:"entry_fee"`
	PrizePool types.Decimal `json:"prize_pool"`
	StartTime string `json:"start_time" binding:"required"`
	EndTime *string `json:"end_time"`
	IsOnline bool `json:"is_online"`
	Location *string `json:"location"`
	RulesText *string `json:"rules_text"`
	SeasonID uint `json:"season_id"`
	OrganizerID uint `json:"organizer_id"`
}

// TournamentUpdateRequest is the PUT/PATCH body — all fields optional.
type TournamentUpdateRequest struct {
	PublicId *string `json:"public_id"`
	Name *string `json:"name"`
	Description *string `json:"description"`
	BracketData *json.RawMessage `json:"bracket_data"`
	Format *TournamentFormatType `json:"format"`
	TournamentType *TournamentTournamentTypeType `json:"tournament_type"`
	MaxPlayers *int `json:"max_players"`
	EntryFee *types.Decimal `json:"entry_fee"`
	PrizePool *types.Decimal `json:"prize_pool"`
	StartTime *string `json:"start_time"`
	EndTime *string `json:"end_time"`
	IsOnline *bool `json:"is_online"`
	Location *string `json:"location"`
	RulesText *string `json:"rules_text"`
	SeasonID *uint `json:"season_id"`
	OrganizerID *uint `json:"organizer_id"`
}

// TournamentResponse is the JSON representation returned by the API.
type TournamentResponse struct {
	ID        uint      `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	PublicId string `json:"public_id"`
	Name string `json:"name"`
	Description *string `json:"description"`
	Status TournamentStatusType `json:"status"`
	BracketData *json.RawMessage `json:"bracket_data"`
	Format TournamentFormatType `json:"format"`
	TournamentType TournamentTournamentTypeType `json:"tournament_type"`
	MaxPlayers int `json:"max_players"`
	EntryFee types.Decimal `json:"entry_fee"`
	PrizePool types.Decimal `json:"prize_pool"`
	StartTime string `json:"startTime"`
	EndTime *string `json:"endTime"`
	IsOnline bool `json:"is_online"`
	Location *string `json:"location"`
	RulesText *string `json:"rules_text"`
	SeasonID uint `json:"season_id"`
	OrganizerID uint `json:"organizer_id"`
}

type Tournament struct {
	gorm.Model
	PublicId string `gorm:"column:public_id;type:varchar(36);not null;uniqueIndex"`
	Name string `gorm:"column:name;not null"`
	Description *string `gorm:"column:description;type:text"`
	Status TournamentStatusType `gorm:"column:status;not null;default:'Draft'"`
	BracketData *json.RawMessage `gorm:"column:bracket_data;type:text"`
	Format TournamentFormatType `gorm:"column:format;not null;default:'Standard'"`
	TournamentType TournamentTournamentTypeType `gorm:"column:tournament_type;not null;default:'Swiss'"`
	MaxPlayers int `gorm:"column:max_players;not null"`
	EntryFee types.Decimal `gorm:"column:entry_fee;type:decimal(10,2);not null;default:0"`
	PrizePool types.Decimal `gorm:"column:prize_pool;type:decimal(10,2);not null;default:0"`
	StartTime string `gorm:"column:start_time;not null"`
	EndTime *string `gorm:"column:end_time"`
	IsOnline bool `gorm:"column:is_online;default:true"`
	Location *string `gorm:"column:location"`
	RulesText *string `gorm:"column:rules_text;type:text"`
	SeasonID uint `gorm:"column:season_id;constraint:OnDelete:RESTRICT"`
	OrganizerID uint `gorm:"column:organizer_id;constraint:OnDelete:RESTRICT"`
	JudgeAssignments []TournamentJudge `gorm:"foreignKey:TournamentID"`
	Registrations []TournamentRegistration `gorm:"foreignKey:TournamentID"`
	Rounds []TournamentRound `gorm:"foreignKey:TournamentID"`
	Prizes []TournamentPrize `gorm:"foreignKey:TournamentID"`
}

func (m *Tournament) ToResponse() TournamentResponse {
	return TournamentResponse{
		ID:        m.ID,
		CreatedAt: m.CreatedAt,
		UpdatedAt: m.UpdatedAt,
		PublicId: m.PublicId,
		Name: m.Name,
		Description: m.Description,
		Status: m.Status,
		BracketData: m.BracketData,
		Format: m.Format,
		TournamentType: m.TournamentType,
		MaxPlayers: m.MaxPlayers,
		EntryFee: m.EntryFee,
		PrizePool: m.PrizePool,
		StartTime: m.StartTime,
		EndTime: m.EndTime,
		IsOnline: m.IsOnline,
		Location: m.Location,
		RulesText: m.RulesText,
		SeasonID: m.SeasonID,
		OrganizerID: m.OrganizerID,
	}
}

func (m *Tournament) ApplyUpdate(req TournamentUpdateRequest) {
	if req.PublicId != nil { m.PublicId = *req.PublicId }
	if req.Name != nil { m.Name = *req.Name }
	if req.Description != nil { m.Description = req.Description }
	if req.BracketData != nil { m.BracketData = req.BracketData }
	if req.Format != nil { m.Format = *req.Format }
	if req.TournamentType != nil { m.TournamentType = *req.TournamentType }
	if req.MaxPlayers != nil { m.MaxPlayers = *req.MaxPlayers }
	if req.EntryFee != nil { m.EntryFee = *req.EntryFee }
	if req.PrizePool != nil { m.PrizePool = *req.PrizePool }
	if req.StartTime != nil { m.StartTime = *req.StartTime }
	if req.EndTime != nil { m.EndTime = req.EndTime }
	if req.IsOnline != nil { m.IsOnline = *req.IsOnline }
	if req.Location != nil { m.Location = req.Location }
	if req.RulesText != nil { m.RulesText = req.RulesText }
	if req.SeasonID != nil { m.SeasonID = *req.SeasonID }
	if req.OrganizerID != nil { m.OrganizerID = *req.OrganizerID }
}

// ── Lifecycle state machine ──────────────────────────────────────
var TournamentAllowedTransitions = map[string]map[string]bool{
		"Draft": {"Registration": true},
		"Registration": {"Ongoing": true, "Cancelled": true},
		"Ongoing": {"Completed": true, "Cancelled": true},
}

func (m *Tournament) AssertTransition(to string) error {
	current := string(m.Status)
	allowed, ok := TournamentAllowedTransitions[current]
	if !ok || !allowed[to] {
		return fmt.Errorf("transition %s -> %s is not allowed", current, to)
	}
	return nil
}

// ── Business operations ──────────────────────────────────────────

func (m *Tournament) Start()  error {
	return fmt.Errorf("Start: not implemented")
}

func (m *Tournament) Cancel()  error {
	return fmt.Errorf("Cancel: not implemented")
}

func (m *Tournament) Complete()  error {
	return fmt.Errorf("Complete: not implemented")
}

func (m *Tournament) GenerateRound()  error {
	return fmt.Errorf("GenerateRound: not implemented")
}

func (m *Tournament) CalculatePrizeDistribution()  (float64, error) {
	return 0.0, fmt.Errorf("CalculatePrizeDistribution: not implemented")
}

func (m *Tournament) RegisterPlayer(playerId int, deckId int)  error {
	return fmt.Errorf("RegisterPlayer: not implemented")
}

func (m *Tournament) IsFull()  (bool, error) {
	return false, fmt.Errorf("IsFull: not implemented")
}

File: axum/src/models/tournaments/tournament.rs
#[derive(Debug, Clone, PartialEq, sqlx::Type, serde::Serialize, serde::Deserialize)]
#[sqlx(type_name = "TEXT", rename_all = "PascalCase")]
pub enum TournamentStatus {
    Draft,
    Registration,
    Ongoing,
    Completed,
    Cancelled,
}

impl TournamentStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            TournamentStatus::Draft => "Draft",
            TournamentStatus::Registration => "Registration",
            TournamentStatus::Ongoing => "Ongoing",
            TournamentStatus::Completed => "Completed",
            TournamentStatus::Cancelled => "Cancelled",
        }
    }
}

impl std::fmt::Display for TournamentStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl std::str::FromStr for TournamentStatus {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Draft" => Ok(TournamentStatus::Draft),
            "Registration" => Ok(TournamentStatus::Registration),
            "Ongoing" => Ok(TournamentStatus::Ongoing),
            "Completed" => Ok(TournamentStatus::Completed),
            "Cancelled" => Ok(TournamentStatus::Cancelled),
            _ => Err(format!("unknown TournamentStatus: {}", s)),
        }
    }
}

impl From<String> for TournamentStatus {
    fn from(s: String) -> Self {
        s.parse().unwrap_or_else(|_| panic!("invalid TournamentStatus: {}", s))
    }
}

#[derive(Debug, Clone, PartialEq, sqlx::Type, serde::Serialize, serde::Deserialize)]
#[sqlx(type_name = "TEXT", rename_all = "PascalCase")]
pub enum TournamentFormat {
    Standard,
    Extended,
    Legacy,
    Vintage,
    Commander,
    Draft,
}

impl TournamentFormat {
    pub fn as_str(&self) -> &'static str {
        match self {
            TournamentFormat::Standard => "Standard",
            TournamentFormat::Extended => "Extended",
            TournamentFormat::Legacy => "Legacy",
            TournamentFormat::Vintage => "Vintage",
            TournamentFormat::Commander => "Commander",
            TournamentFormat::Draft => "Draft",
        }
    }
}

impl std::fmt::Display for TournamentFormat {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl std::str::FromStr for TournamentFormat {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Standard" => Ok(TournamentFormat::Standard),
            "Extended" => Ok(TournamentFormat::Extended),
            "Legacy" => Ok(TournamentFormat::Legacy),
            "Vintage" => Ok(TournamentFormat::Vintage),
            "Commander" => Ok(TournamentFormat::Commander),
            "Draft" => Ok(TournamentFormat::Draft),
            _ => Err(format!("unknown TournamentFormat: {}", s)),
        }
    }
}

impl From<String> for TournamentFormat {
    fn from(s: String) -> Self {
        s.parse().unwrap_or_else(|_| panic!("invalid TournamentFormat: {}", s))
    }
}

#[derive(Debug, Clone, PartialEq, sqlx::Type, serde::Serialize, serde::Deserialize)]
#[sqlx(type_name = "TEXT", rename_all = "PascalCase")]
pub enum TournamentTournamentType {
    Swiss,
    SingleElimination,
    DoubleElimination,
    RoundRobin,
}

impl TournamentTournamentType {
    pub fn as_str(&self) -> &'static str {
        match self {
            TournamentTournamentType::Swiss => "Swiss",
            TournamentTournamentType::SingleElimination => "SingleElimination",
            TournamentTournamentType::DoubleElimination => "DoubleElimination",
            TournamentTournamentType::RoundRobin => "RoundRobin",
        }
    }
}

impl std::fmt::Display for TournamentTournamentType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl std::str::FromStr for TournamentTournamentType {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Swiss" => Ok(TournamentTournamentType::Swiss),
            "SingleElimination" => Ok(TournamentTournamentType::SingleElimination),
            "DoubleElimination" => Ok(TournamentTournamentType::DoubleElimination),
            "RoundRobin" => Ok(TournamentTournamentType::RoundRobin),
            _ => Err(format!("unknown TournamentTournamentType: {}", s)),
        }
    }
}

impl From<String> for TournamentTournamentType {
    fn from(s: String) -> Self {
        s.parse().unwrap_or_else(|_| panic!("invalid TournamentTournamentType: {}", s))
    }
}

#[derive(Debug, Clone, sqlx::FromRow, serde::Serialize)]
pub struct Tournament {
    pub id: i64,
    pub created_at: String,
    pub updated_at: String,
    pub name: String,
    pub description: Option<String>,
    pub status: TournamentStatus,
    pub format: TournamentFormat,
    pub tournament_type: TournamentTournamentType,
    pub max_players: i64,
    pub entry_fee: f64,
    pub prize_pool: f64,
    #[serde(rename = "startTime")]
    pub start_time: String,
    #[serde(rename = "endTime")]
    pub end_time: Option<String>,
    #[serde(serialize_with = "crate::serde_utils::serialize_i64_as_bool")]
    pub is_online: i64,
    pub location: Option<String>,
    pub rules_text: Option<String>,
    pub season_id: i64,
    pub organizer_id: i64,
}

#[derive(Debug, serde::Deserialize)]
pub struct TournamentCreateRequest {
    pub name: String,
    pub description: Option<String>,
    pub status: TournamentStatus,
    pub format: TournamentFormat,
    pub tournament_type: TournamentTournamentType,
    pub max_players: i64,
    pub entry_fee: f64,
    pub prize_pool: f64,
    pub start_time: String,
    pub end_time: Option<String>,
    pub is_online: bool,
    pub location: Option<String>,
    pub rules_text: Option<String>,
    pub season_id: i64,
    pub organizer_id: i64,
}

#[derive(Debug, serde::Deserialize)]
pub struct TournamentUpdateRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub status: Option<TournamentStatus>,
    pub format: Option<TournamentFormat>,
    pub tournament_type: Option<TournamentTournamentType>,
    pub max_players: Option<i64>,
    pub entry_fee: Option<f64>,
    pub prize_pool: Option<f64>,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
    pub is_online: Option<bool>,
    pub location: Option<String>,
    pub rules_text: Option<String>,
    pub season_id: Option<i64>,
    pub organizer_id: Option<i64>,
}

impl Tournament {
    pub fn assert_transition(&self, to: &str) -> Result<(), String> {
        let allowed: &[(&str, &str)] = &[
            ("Draft", "Registration"),
            ("Registration", "Ongoing"),
            ("Registration", "Cancelled"),
            ("Ongoing", "Completed"),
            ("Ongoing", "Cancelled"),
        ];
        let current = &self.status;
        if allowed.iter().any(|(from, t)| *from == current.as_str() && *t == to) {
            Ok(())
        } else {
            Err(format!("transition {} -> {} not allowed", current, to))
        }
    }
}
